### PR TITLE
adding type hints

### PR DESCRIPTION
**Pull request by Phil-hacker**
PR for #43

### CODE REVIEW

**Comment by Phil-hacker** [comment]
i changed the constructor for DiscriminatingDictionary for better type hints
DiscriminatingDictionary should probably be refactored to allow for better type hints

File: litemapy/storage.py [hunk 1]
```diff
@@ -74,29 +78,32 @@ def __setitem__(self, index, value):
...
-    def __contains__(self, value):
+    def __contains__(self, value: int) -> bool:
         for v in self:
             if v == value:
                 return True
         return False
 
 
 class DiscriminatingDictionary(dict):
+    validator: Callable[[Any],tuple[bool, str]]
```
Comment (SmylerMC) [inline litemapy/storage.py:102]: `validator` gets both the key and its value as parameters.

```suggestion
    validator: Callable[[Any, Any],tuple[bool, str]]
```

File: litemapy/storage.py [hunk 2]
```diff
@@ -74,29 +78,32 @@ def __setitem__(self, index, value):
...
+    on_add: Callable[[Any,Any],None] | None
+    on_remove: Callable[[Any,Any],None] | None
 
-    def __init__(self, validator, *args, **options):
+    def __init__(self, validator: Callable[[Any],tuple[bool, str]], *args, onadd: Callable[[Any,Any],None] | None = None,onremove: Callable[[Any,Any],None] | None = None, **options) -> None:
```
Comment (SmylerMC) [inline litemapy/storage.py:106]: This line is way too long and has formatting problems

File: litemapy/storage.py [hunk 3]
```diff
@@ -150,30 +157,30 @@ def setdefault(self, key, *args):
             self.__on_add(key, default)
         return r
 
-    def update(self, other):
+    def update(self, other: 'DiscriminatingDictionary') -> None:
```
Comment (SmylerMC) [inline litemapy/storage.py:160]: This should accept vanilla dictionaries as well.

File: litemapy/boxes.py [hunk 4]
```diff
@@ -17,5 +17,5 @@ def block_is_in_box(block, box):
     return x_min <= x <= x_max and y_min <= y <= y_max and z_min <= z <= z_max
 
 
-def box_is_in_box(box1, box2):
+def box_is_in_box(box1: tuple[tuple[int, int, int], tuple[int, int, int]], box2: tuple[tuple[int, int, int], tuple[int, int, int]]) -> bool:
```
Comment (SmylerMC) [inline litemapy/boxes.py:20]: Line too long.

**Review by SmylerMC — CHANGES_REQUESTED** [review]
Pretty good overall with a few problems:
- The binary OR syntax for types was added in Python 3.10. We need to support 3.9 so I think the only solution is to fallback to [`typing.Union`](https://docs.python.org/3/library/typing.html#typing.Union). This is preventing CI to pass.
- Formatting (I can fix that myself before merging if you prefer):
  - classes and function on the global scope need two blank lines prior to their declaration,
  - Methods need one blank line prior to their declaration,
  - `=` in keyword arguments need spaces on both sides when used with annotations
  - Comma should be followed by a blank.

**Review by SmylerMC — CHANGES_REQUESTED** [review]
Still has a few formatting problems, I taking care of it

**Review by SmylerMC — APPROVED** [review]
Looks all good now.

**Comment by SmylerMC** [comment]
Merging, thank you @Phil-hacker !

**Pull request merged**